central/complianceoperatorintegration/datastore/datastore_impl.go

```diff
@@ -0,0 +1,127 @@
+	if ok, err := integrationSAC.ReadAllowed(ctx); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, nil
+	}
+
+	return ds.storage.GetByQuery(ctx, search.NewQueryBuilder().
+		AddStrings(search.ClusterID, clusterID).ProtoQuery())
+}
+
+// AddComplianceIntegration is pass-through to the underlying store.
+func (ds *datastoreImpl) AddComplianceIntegration(ctx context.Context, integration *storage.ComplianceIntegration) (string, error) {
+	if ok, err := integrationSAC.WriteAllowed(ctx); err != nil {
+		return "", err
+	} else if !ok {
+		return "", sac.ErrResourceAccessDenied
+	}
+
+	if integration.Id == "" {
+		integrationID, err := ds.getIntegrationID(ctx, integration.GetClusterId())
+		if err != nil {
+			return "", err
+		}
+
+		integration.Id = integrationID
+	}
+
+	err := ds.storage.Upsert(ctx, integration)
+	if err != nil {
+		return "", err
+	}
+	return integration.Id, nil
+}
+
+// UpdateComplianceIntegration is pass-through to the underlying store.
+func (ds *datastoreImpl) UpdateComplianceIntegration(ctx context.Context, integration *storage.ComplianceIntegration) error {
+	if ok, err := integrationSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	if integration.Id == "" {
+		return errors.New("Unable to update compliance integration without an ID")
+	}
+
+	return ds.storage.Upsert(ctx, integration)
+}
+
+// RemoveComplianceIntegration is pass-through to the underlying store.
+func (ds *datastoreImpl) RemoveComplianceIntegration(ctx context.Context, id string) error {
+	if ok, err := integrationSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+	return ds.storage.Delete(ctx, id)
+}
+
+// RemoveComplianceIntegrationByCluster removes all the compliance integrations for a cluster
+func (ds *datastoreImpl) RemoveComplianceIntegrationByCluster(ctx context.Context, clusterID string) error {
+	if ok, err := integrationSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	return ds.storage.DeleteByQuery(ctx, search.NewQueryBuilder().
+		AddStrings(search.ClusterID, clusterID).ProtoQuery())
+}
+
+func (ds *datastoreImpl) getIntegrationID(ctx context.Context, clusterID string) (string, error) {
+	var integrationID string
+
+	existingIntegrations, err := ds.storage.GetByQuery(ctx, search.NewQueryBuilder().
+		AddStrings(search.ClusterID, clusterID).ProtoQuery())
+	if err != nil {
+		return "", err
+	}
+	// Due to unique constraint on cluster/namespace, should only return 1
+	if len(existingIntegrations) > 1 {
+		return "", errors.Wrapf(err, "Multiple integrations already registered to cluster %q", clusterID)
+	}
+
+	if len(existingIntegrations) == 1 {
+		integrationID = existingIntegrations[0].Id
+	} else {
+		integrationID = uuid.NewV4().String()
+	}
```

Comment:
This pattern for add obj path is incorrect. This should not be handled in the datastore but by the pipeline manager.

Reply:
Yeah I went back and forth on that.  I will remove this logic.  It will re-appear later when I implement the pipeline manager.